Expanding "shield" functionality (example uses WNC-M14A2A and SARA modem)

mike-scott opened the pull request:
I took some time today to see if I could setup a "shield" to represent the AT&T LTE-M WNC-M14A2A-based Arduino-R3 compatible shield.

This is very rough work and generates a few warnings when compiling, but this RFC patch series handles the following "special cases" which I encountered:
- I felt like a shield-based .conf file simplified everything.  So I added handling for that (as opposed to adding overlay-conf files into each sample or creating special samples for the shield).
- AT&T LTE-M shield uses non-standard UART pins on the headers and Nordic UART needs tx, rx, cts and rts pins configured for nRF52840_pca10056, so I added handling for [shield]_X_[board].overlay files which could be board specific.
- No current shields use GPIOs, and in order to configure the A0-A5 & D0-D13 Arduino pins, I created a new dts binding for arduino-header-r3.
- Then I configured those pins in the frdm_k64f and nrf52840_pca10056 DTS files and assign them arduino_d1/d2 handle names.
- I adjusted the WNC-M14A2A modem driver and DTS bindings slightly to allow for the use of aliases for the pins instead of the dts-binding gpios.  This way driver can be used for the shield-based WNC-M14A2A or if a piece of HW had the modem directly connected, it could still be configured completely in the board's DTS.
- Finally I add the actual wnc_m14a2a shield complete with it's own overlay file, conf file, and board-based overlays.  (Missing docs for the moment).

The following now works to build the LwM2M client for each of the 2 configured boards:
```
$ cd samples/net/lwm2m_client/
$ mkdir build && cd build
$ cmake -DBOARD=frdm_k64f -DSHIELD=wnc_m14a2a ..
or
$ cmake -DBOARD=nrf52840_pca10056 -DSHIELD=wnc_m14a2a ..
```

These warnings are generated since I named the alias "gpios":
```
nrf52840_pca10056.dts_compiled: Warning (gpios_property): /aliases:dt-wnc-m14a2a-0-mdm-boot-mode-sel-gpios: cell 0 is not a phandle reference
nrf52840_pca10056.dts_compiled: Warning (gpios_property): /aliases:dt-wnc-m14a2a-0-mdm-boot-mode-sel-gpios: Could not get phandle node for (cell 0)
nrf52840_pca10056.dts_compiled: Warning (gpios_property): /aliases:dt-wnc-m14a2a-0-mdm-power-gpios: cell 0 is not a phandle reference
nrf52840_pca10056.dts_compiled: Warning (gpios_property): /aliases:dt-wnc-m14a2a-0-mdm-power-gpios: Could not get phandle node for (cell 0)
nrf52840_pca10056.dts_compiled: Warning (gpios_property): /aliases:dt-wnc-m14a2a-0-mdm-keep-awake-gpios: cell 0 is not a phandle reference
nrf52840_pca10056.dts_compiled: Warning (gpios_property): /aliases:dt-wnc-m14a2a-0-mdm-keep-awake-gpios: Could not get phandle node for (cell 0)
nrf52840_pca10056.dts_compiled: Warning (gpios_property): /aliases:dt-wnc-m14a2a-0-mdm-reset-gpios: cell 0 is not a phandle reference
nrf52840_pca10056.dts_compiled: Warning (gpios_property): /aliases:dt-wnc-m14a2a-0-mdm-reset-gpios: Could not get phandle node for (cell 0)
nrf52840_pca10056.dts_compiled: Warning (gpios_property): /aliases:dt-wnc-m14a2a-0-mdm-shld-trans-ena-gpios: cell 0 is not a phandle reference
nrf52840_pca10056.dts_compiled: Warning (gpios_property): /aliases:dt-wnc-m14a2a-0-mdm-shld-trans-ena-gpios: Could not get phandle node for (cell 0)
```

However, it looks like the generated include file has the right defines/values in it.

I am also working on the U-blox SARA R41xx modem driver, which is present in the Particle.io Boron board as well as a SparkFun LTE-M Arduino shield.  It uses the standard UART exposed by Arduino so it won't need per board .overlay files, but this series solves a lot of the issues I am hitting with that driver as well (reset and power GPIOs, etc).

Please discuss.

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -308,6 +321,50 @@ foreach(root ${BOARD_ROOT})
   endif()
+
+  # Match the .conf files in the shield directories to make sure we are
+  # finding shields, e.g. wnc_m14a2a/wnc_m14a2a.conf.conf
```
Same below.

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -289,6 +289,19 @@ foreach(root ${BOARD_ROOT})
+          shield_dts_files
+          ${shield_dir}/${s_path}
+        )
+      endif()
```
The same applies to the shield .conf files.

mike-scott replied:
Moved to this naming convention.

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -289,6 +289,19 @@ foreach(root ${BOARD_ROOT})
 
   if(DEFINED SHIELD)
     foreach(s ${SHIELD_AS_LIST})
+      # search list for shield_X_board name
```
The same applies to the shield .conf files (I believe).

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -308,6 +321,50 @@ foreach(root ${BOARD_ROOT})
+
+  # Match the .conf files in the shield directories to make sure we are
+  # finding shields, e.g. wnc_m14a2a/wnc_m14a2a.conf.conf
+  file(GLOB_RECURSE shields_conf_refs_list
```
I don't think we need this list, I believe that since we know which shields exist we can just check, for each enabled shield, if a corresponding .conf file exists, and if so add it to the list of shield_conf_files.

SebastianBoe commented on cmake/kconfig.cmake:
```diff
@@ -89,6 +89,7 @@ set(
   ${BOARD_DEFCONFIG}
   ${CONF_FILE_AS_LIST}
   ${OVERLAY_CONFIG_AS_LIST}
+  ${shield_conf_files}
```
I'd think shield_conf_files were at a similar priority to BOARD_DEFCONFIG, so after BOARD_DEFCONFIG.

mike-scott replied:
Yes, you are correct.  Fixed.

SebastianBoe reviewed (changes requested):
The approach looks sane.

I'd prefer it if we could express this pattern in DT source code, instead of in the toolchain, but AFAICT this is impossible.

This pattern needs documentation.

mike-scott commented:
@SebastianBoe Thank you for the great feedback!  I'll put up a v2 with your comments accounted for.

erwango commented on drivers/modem/Kconfig.wncm14a2a:
```diff
@@ -11,6 +11,7 @@ menuconfig MODEM_WNCM14A2A
 	depends on MODEM
 	select MODEM_RECEIVER
 	select NET_OFFLOAD
+	select GPIO
```
Can't we clean up few of the options below now?

mike-scott replied:
Cleaned up the HW specific options.

mike-scott commented on boards/shields/wnc_m14a2a/wnc_m14a2a.conf:
```diff
@@ -2,14 +2,11 @@
 CONFIG_MODEM=y
```
Fixed sample.yaml.  Nice catch.

mike-scott commented on cmake/app/boilerplate.cmake:
```diff
@@ -279,7 +279,7 @@ foreach(root ${BOARD_ROOT})
-  # x_nucleo_iks01a1/x_nucleo_iks01a1.overlay;x_nucleo_iks01a2/x_nucleo_iks01a2.overlay
+  # wnc_m14a2a/wnc_m14a2a.overlay;wnc_m14a2a/wnc_m14a2a-nrf52840_pca10056.overlay
   # we construct a list of shield names by extracting file name and
   # removing the extension.
```
Went with your recommended approach of keeping the shield list as it is and using EXIST to add the files where appropriate.  Much cleaner.

mike-scott commented on boards/shields/wnc_m14a2a/wnc_m14a2a_X_frdm_k64f.overlay:
```diff
@@ -0,0 +1,23 @@
+		label = "wncm14a2a";
+		status = "ok";
+
+		/* GPIOs are set by alias above */
```
I added a comment explaining the aliases in the main wnc_m14a2a.overlay file.  This was a left over reference to those aliases and was removed in the new PR.

mike-scott commented:
Fixed gitlint issue

mike-scott commented:
@erwango Are you planning to be main CODEOWNER for boards/shields/?

mike-scott commented on boards/shields/wnc_m14a2a/wnc_m14a2a.overlay:
```diff
@@ -0,0 +1,30 @@
+ */
+
+/ {
+	/*
```
This needs to be discussed as I think a change in our DTS processing is needed for:
1) To remove the warning generated by these aliases using labels with a "gpios" portion in their name
OR
2) Perhaps we can make a change which would allow the actual GPIO setting in the modem DTS to be assigned a label:
mdm-boot-mode-sel-gpios = &arduino_d1;
This would make these hacky aliases go away.

erwango replied:
I'll try to discuss this at connect with dts guys.

erwango commented:
@mike-scott 

> Are you planning to be main CODEOWNER for boards/shields/?

I thought I was already, but indeed, it appears this is not the case. So, if you volunteer, please go ahead. Otherwise, answer is yes.

MaureenHelm commented on boards/shields/wnc_m14a2a/boards/frdm_k64f.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_UART_MCUX_2=y
```
What if instead we add a CONFIG_ARDUINO_SERIAL to `drivers/serial/Kconfig`, and then the boards can intelligently enable the right instance like they do with the console uart? Something like this in `boards/arm/frdm_k64f/Kconfig.defconfig`

```
if UART_MCUX

config UART_MCUX_0
        default y if UART_CONSOLE
config UART_MCUX_2
        default y if ARDUINO_SERIAL
config UART_MCUX_3
        default y if BT_UART

endif # UART_MCUX
```

mike-scott replied:
@MaureenHelm I like this idea.  I was struggling with the abstraction between a shield's UART (arduino_serial) and a HW specific UART (Like if the driver needed to be used directly).

A great example of this is the Particle.io Boron board which has a U-blox SARA-R41xx modem connected to the nRF52840 directly via UART1 vs. a SparkFun LTE Cat M1 shield which has the U-blox SARA-R41xx connected to Arduino R3-compatible headers.

The shield configuration should only reference the arduino_serial and stay away from anything too specific to the device.

However, this approach won't work for the WNC-M14A2A modem.   The UART pins are on non-standard Arduino pins.  The traditional Arduino UART is actually C16/C17 pins (UART_3) for K64F.

Technically, ARDUINO_SERIAL should enable UART_3 (where BT_UART is identified).   And, the BT_UART setting could go away if we migrate that to a shield setup.

MaureenHelm reviewed (commented):
Thanks @mike-scott for tackling this problem! It's been blocking me from converting the MCR20A to a shield for quite some time.

mike-scott commented:
> @mike-scott
> 
> > Are you planning to be main CODEOWNER for boards/shields/?
> 
> I thought I was already, but indeed, it appears this is not the case. So, if you volunteer, please go ahead. Otherwise, answer is yes.

@erwango Feel free to take ownership of the boards/shields dir.  My Zephyr plate is fairly full considering I'm working at a starving startup. :)

mike-scott commented:
recheck

mike-scott commented:
Looks like at least 1 remaining issue would be to establish CODEOWNER for:
dts/bindings/gpio/arduino-header-r3.yaml

@erwango Would you want to manage these shield related bindings as a part of being owner for boards/shields?  Not sure if you took a close look at how I implemented the binding or not.

mike-scott commented:
Updated with the following changes:
- Split out the Arduino header additions into separate patches by board: frdm_k64f and nrf52840_pca10056.
- Added a new Arduino header addition patch for disco_l475_iot1 board.
- Rebased

mike-scott commented:
Still the CODEOWNER question about dts/bindings/gpio/arduino-header-r3.yaml

@erwango I can probably own that file.

mike-scott commented:
Question for all: should it named arduino-header-r3.yaml or perhaps something more generic as it doesn't have to be used for Arduino R3 specific uses.  It merely allows GPIOs to be exposed for DTS use which are not SoC specific.

erwango commented on boards/arm/disco_l475_iot1/disco_l475_iot1.dts:
```diff
@@ -45,6 +45,76 @@
 		eswifi0 = &wifi0;
 	};
+
+	header_pins {
```
What about pulling all arduino definitions into a disco_l475_iot1_arduino.dtsi (at current location) than would be included in disco_l475_iot1.dts?

mike-scott replied:
I can do that.

erwango commented on cmake/app/boilerplate.cmake:
```diff
@@ -306,6 +306,33 @@ foreach(root ${BOARD_ROOT})
+      endif()
+
+      # search for shield/shield.conf file
+      if(EXISTS ${shield_dir}/${s}/${s}.conf)
```
I think shields should follow the same rule as boards, ie configure the HW, but not activate features, which is application responsibility. So in general, having Kconfig.defconfig (conditional confguration), rather than .conf or def_config that set Kconfig symbols.

mike-scott replied:
I don't agree entirely.  The user has explicitly added the SHIELD configuration to the build.  In my head this is exactly like specifying overlay-*.conf (with the added benefit that SHIELD configuration is accessible across all samples).

The Kconfig.defconfig file is processed **every time** the board is built for regardless of the configuration.  Hence the conditional nature.  This is not the same IMHO.

I *do* agree WNC-M14A2A .conf files are a bit messy:
1) Settings like this are fine:
```
# Enable WNC-M14A2A LTE-M Modem
CONFIG_MODEM=y
CONFIG_MODEM_SHELL=y
CONFIG_MODEM_WNCM14A2A=y
CONFIG_UART_INTERRUPT_DRIVEN=y
```
2) The following are definitely hackish.  They are specific to the transport medium and driver, not necessarily the shield:
```
# extend retry timing to 20 seconds for LTE/LTE-M
CONFIG_COAP_INIT_ACK_TIMEOUT_MS=20000

# Hack: disable IPv6 for now
CONFIG_NET_IPV6=n
CONFIG_NET_CONFIG_NEED_IPV6=n
```

I'm looking to fix these in a different way in the future.

3) Most SHIELD setups won't need the boards/* additions like `frdm_k64f.conf` which sets `CONFIG_UART_MCUX_2=y`.  This is due to the serial port being on a non-standard Arduino header pin (standard serial is `CONFIG_UART_MCUX_3`).
Most of the settings in `nrf52840_pca10056.conf` can probably be abstracted like @MaureenHelm mentioned where the board sets up the following if `CONFIG_ARDUINO_SERIAL` is enabled:
```
CONFIG_GPIO_NRF_P1=y
CONFIG_UART_1_NRF_UARTE=y
```

Not sure what to do about the: `CONFIG_UART_1_NRF_FLOW_CONTROL=y` setting.  This particular modem (driver) needs it and it's not a setting controlled by DTS.  This one might need to stay in `boards/nrf52840_pca10056.conf`.

erwango replied:
> I don't agree entirely. The user has explicitly added the SHIELD configuration to the build. In my head this is exactly like specifying overlay-*.conf (with the added benefit that SHIELD configuration is accessible across all samples).

OTOH, we can have boards with the same HW component (be it a modem, a sensor, ..) out of the box. From application point of view, to use this HW component, being direct part of the BSP or added as a shield, same configuration should be done (MODEM, INTERRUPT_DRIVEN, ..).
So shield configuration should add the missing configuration bits between application and the board default one.

> The Kconfig.defconfig file is processed every time the board is built for regardless of the configuration. Hence the conditional nature. This is not the same IMHO.

I've missed your point here.

mike-scott replied:
@erwango 

> I've missed your point here.

I was trying to distinguish between a file that would only be processed if the user includes it on the build line vs. Kconfig.defconfig which is always processed.  Can ignore see below:

I added 1 more commit as an example for enabling the SARA-R4 modem on the Particle.io Boron board here: https://github.com/zephyrproject-rtos/zephyr/pull/14057/commits/40321361e59c6364fde089b914221b14258a51fe
(Boron has built-in modem)

This is a concrete example we can discuss as it uses the more traditional method of Kconfig.defconfig: if CONFIG_MODEM is enabled, it chooses the SARA_R4 driver.  The obvious problem with the Boron board is that it has several methods of communication:
- BLE 6lowpan (via CONFIG_NET_L2_BT)
- 802.15.4 via (via CONFIG_NET_L2_IEEE802154 or CONFIG_NET_L2_OPENTHREAD)
- SARA-R4 modem (via CONFIG_MODEM)

The only way Zephyr currently has of dealing with these is to add overlay-*.conf files into each sample:
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/echo_client/overlay-bt.conf
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/echo_client/overlay-802154.conf
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/echo_client/overlay-ot.conf

Such overlay files are typically carbon copies of each other across all of the samples and they aren't scaling well.  Consider these 2 overlay-bt.conf files:
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/lwm2m_client/overlay-bt.conf
and
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/sockets/echo_client/overlay-bt.conf

In a perfect world we have a `boards/common-overlays` dir and this is part of the overlay-*.conf search path when checking -DOVERLAY_CONFIG= files.   This is something we could do regardless of shield-based changes.

Let's also take a look at the current contents of:
https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/net/lwm2m_client/overlay-wncm14a2a.conf

This is incredibly similar to the wnc_m14a2a.conf file added under boards/shields/wnc_m14a2a/.

Let's examine your proposal of doing the settings as conditional in a Kconfig.defconfig file:
- Move not only the shield's .conf file into Kconfig.defconfig format, but also create conditional handling for all of the supported boards as well (if there are boards/*.conf) because there are dependencies in those files which might or might not be enabled.
- After that's done, the user would still need to add -DOVERLAY_CONFIG=* with a generic opt-in overlay file which enables the main piece of hardware / functionality.

This doesn't *seem* like a better solution unless I'm missing an example where the user would specifically add -DSHIELD=* to his build line, but **NOT** want the functionality of the shield enabled.   Perhaps you have a good example in mind?

erwango commented on boards/shields/wnc_m14a2a/boards/nrf52840_pca10056.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_GPIO_NRF_P1=y
```
Rather do it as Kconfig.defconfig than .conf direct setting style

mike-scott replied:
Like I said above, this particular setting can be abstracted under `CONFIG_ARDUINO_SERIAL` and handled in the board's Kconfig.defconfig.  The nRF52840's don't have the same problem as K64F because they can set any pin to serial.   The tricky part here is that these settings in the .overlay are unique to this shield:
```
	tx-pin = <46>;
	rx-pin = <45>;
	rts-pin = <44>;
	cts-pin = <47>;
```

The standard Arduino serial pins are different.

mike-scott commented:
The last few days, I've been working on a more "standard" modem shield: the SparkFun LTE Cat M1 / NB-IoT U-Blox SARA-R410M shield:
https://www.sparkfun.com/products/14997

In theory, this shield should "just work" with **any** of our boards that have Arduino-R3 compatible headers.  That list is quite long now.

I will probably stack the SARA-R4 enablement patches on top of this PR so that we can see what a "good" shield looks like and how it can be abstracted in the board files.  I haven't done that abstraction yet, but right now the branch is here: https://github.com/mike-scott/zephyr/commits/master-sara-modem

This new modem driver is a great example for another reason: it's also in the Particle.IO Boron board as a directly connected IC.  So it needs to work as `-DSHIELD=sara_r41xx` using the `CONFIG_ARDUINO_SERIAL` abstraction or when building for `-DBOARD=particle_boron` with `-DCONFIG_MODEM=y` (eventually the board's Kconfig.defconfig will conditionally enable the modem driver if `CONFIG_MODEM=y`)

mike-scott commented:
For discussion, this is the commit adding the shield config for sara_r41xx:
https://github.com/zephyrproject-rtos/zephyr/pull/14057/commits/b7ce8162854d4e44e178c84bd12eeb014314574b

mike-scott commented:
It's worth noting that the SARA-R4 addition relies on this: https://github.com/zephyrproject-rtos/zephyr/pull/13618 build will fail otherwise.

mike-scott commented:
Changes:
- Rebased to include NET_OFFLOAD fixes in master.  This should now work for testing.
- Minor driver changes for SARA-R4:
  - Driver renamed to: ublox_sara_r4
  - Shield renamed to: sparkfun_sara_r4

Still TODO:
- Abstract shield serial port as ARDUINO_SERIAL
- Possibly rename arduino-header-r3 binding to something more generic for shield pins
- Ongoing discussion with @erwango regarding Kconfig.defconfig conditional settings vs. overlay-*.conf style settings. (PROs vs. CONs)

mike-scott commented:
As an example, I use this to build the LwM2M client for the SparkFun SARA-R4 shield w/ AT&T IoT/Jasper SIM (using IP for leshan.eclipse.org LwM2M server):
```
west build -d build/lwm2m_client -b frdm_k64f -s zephyr/samples/net/lwm2m_client -- \
  -DSHIELD=sparkfun_sara_r4 -DCONFIG_NET_CONFIG_PEER_IPV4_ADDR=\"5.39.83.206\" \
  -DCONFIG_MODEM_UBLOX_SARA_R4_APN=\"m2m.com.attz\"
```

erwango commented:
> Possibly rename arduino-header-r3 binding to something more generic for shield pins

What about shield-connector-arduino? (might be worth keeping the r3, though)

> Ongoing discussion with @erwango regarding Kconfig.defconfig conditional settings vs. overlay-*.conf style settings. (PROs vs. CONs)

Yes, processing that, will come back to you soon.

WilliamGFish commented:
@mike-scott 
Continuing with the Driver / Shield discussion the Particle.io boards additionally use the Espressif ESP32 on their Argon board.

The ESP32/8266 is a often used shield in the Arduino world as a way to add WiFi functionality, so follows the same logic you applied above.

I would lean towards having a device driver and a shield overlay config. Just my humble opinion.

When you have settled on a standard in would be looking at the ESP8266 driver that was started (AT Command driven) which could follow the same standards.
Add wifi offload driver for ESP8266 #10539
https://github.com/zephyrproject-rtos/zephyr/pull/10539/commits/338144c3b79481204599cc8493a07fbe22ecaa4c

erwango commented:
@mike-scott, good news: there is an upcoming dt specification that defines the way to write connectors/pin_headers:
https://github.com/devicetree-org/devicetree-specification/blob/4b1dac80eaca45b4babf5299452a951008a5d864/source/devicetree-basics.rst#nexus-nodes-and-specifier-mapping

Good part is that it provide a way to define a pin-map and adds the possibility to define GPIO flag in the shield overlay, which was missing. Though it requires some work on our dts extraction script to walk the last mile of the phandle resolution.

So basically, on board side, we'll define header as follows:
```
	arduino_header: connector {
	        #gpio-cells = <2>;
	        gpio-map = <0 0 &gpiob 2 0>,
	                   <1 0 &gpiob 3 0>,
	                   <2 0 &gpiob 10 0>,
	                   <3 0 &gpiob 11 0>,
	                   <4 0 &gpioc 11 0>,
	                   <5 0 &gpioc 10 0>,
	                   <6 0 &gpioc 16 0>,
	                   <7 0 &gpioc 17 0>,
	                   <8 0 &gpiob 9 0>,
	                   <9 0 &gpioa 1 0>,
	                   <10 0 &gpiob 23 0>,
	                   <11 0 &gpioa 2 0>,
	                   <12 0 &gpioc 2 0>,
	                   <13 0 &gpioc 3 0>,
	                   <14 0 &gpioc 12 0>,
	                   <15 0 &gpioc 4 0>,
	                   <16 0 &gpiod 0 0>,
	                   <17 0 &gpiod 2 0>,
	                   <18 0 &gpiod 3 0>,
	                   <19 0 &gpiod 1 0>,
	                   <20 0 &gpioe 25 0>,
	                   <21 0 &gpioe 24 0>;
	        gpio-map-mask = <0xf 0x0>;
	        gpio-map-pass-thru = <0x0 0x1>;
	};
```
And, in shield overlay:
```
   	expansion_device {
   	        ublox-sara-r4-0-mdm-reset-gpios = <&arduino_header 4 0>;
   	        ublox-sara-r4-0-mdm-power-gpios = <&arduino_header 5 0>;
   	};
```
Processed by dtc (current zephyr supported version), this ends up in the following:
```
	arduino_header: connector {
		#gpio-cells = <0x02>;
		gpio-map = <0x00 0x00 0x06 0x02 0x00>, <0x01 0x00 0x06 0x03 0x00>, <0x02 0x00 0x06 0x0a 0x00>, <0x03 0x00 0x06 0x0b 0x00>, <0x04 0x00 0x03 0x0b 0x00>, <0x05 0x00 0x03 0x0a 0x00>, <0x06 0x00 0x03 0x10 0x00>, <0x07 0x00 0x03 0x11 0x00>, <0x08 0x00 0x06 0x09 0x00>, <0x09 0x00 0x07 0x01 0x00>, <0x0a 0x00 0x06 0x17 0x00>, <0x0b 0x00 0x07 0x02 0x00>, <0x0c 0x00 0x03 0x02 0x00>, <0x0d 0x00 0x03 0x03 0x00>, <0x0e 0x00 0x03 0x0c 0x00>, <0x0f 0x00 0x03 0x04 0x00>, <0x10 0x00 0x09 0x00 0x00>, <0x11 0x00 0x09 0x02 0x00>, <0x12 0x00 0x09 0x03 0x00>, <0x13 0x00 0x09 0x01 0x00>, <0x14 0x00 0x08 0x19 0x00>, <0x15 0x00 0x08 0x18 0x00>;
		gpio-map-mask = <0x0f 0x00>;
		gpio-map-pass-thru = <0x00 0x01>;
		phandle = <0x0a>;
	};

	expansion_device {
		ublox-sara-r4-0-mdm-reset-gpios = <0x0a 0x04 0x00>;
		ublox-sara-r4-0-mdm-power-gpios = <0x0a 0x05 0x00>;
	};
```
As you can see, at this step `ublox-sara-r4-0-mdm-reset-gpios` as a phandle that redirect to `arduino_header` but not directly the gpio phandle we expect. There are some operations to be done with `gpio-map-mask` and `gpio-map-pass-thru`. These are described in device tree specification and have to be implemented in `scripts/dts/extract_dts_includes.py` so we can generate whatever #defines 
we need.

So far I haven't dig much more and this might have some limitation in some cases.
At least, it allows to compile the dt source in a clean way and exit with a specified output. As a next step I'll have a try to hack the extract dts script to see how much work we still have ahead of us. But if this is zephyr work only, we should be able to sort this out in a reasonable time frame.

avisconti commented:
This work is excellent and permits to handle complex examples of shields.
I need this work to get done in order to clean up my x_nucleo_cca02m1 microphones shield stuff.

weinholtendian commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -0,0 +1,1854 @@
+	ARG_UNUSED(dev);
+
+	/* check for valid pinconfig */
+	__ASSERT(sizeof(pinconfig) == MAX_MDM_CONTROL_PINS,
```
Shouldn't `sizeof(pinconfig)` be `ARRAY_SIZE(pinconfig)` here? The size of pinconfig is larger than the number of pins because it's an array with a struct per pin.

mike-scott replied:
Huh.  I suppose I never turned on asserts and saw this fail.  But you are correct.

weinholtendian commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -0,0 +1,1854 @@
+		LOG_ERR("send_data error: %d", ret);
+	}
+
+	net_pkt_unref(pkt);
```
Correct me if I'm wrong, but will not the caller do `net_pkt_unref` here if `ret < 0`? So this should be in an else branch. I'm using DTLS and saw the error `<err> net_pkt: *** ERROR *** pkt 0x20027e78 is freed already by offload_sendto():1717 (context_sendto():1473)`.

mike-scott replied:
Also correct here.  Fixing.

mike-scott commented:
Rebasing on current master.  And @weinholtendian changes look like this:
```
diff --git a/drivers/modem/ublox-sara-r4.c b/drivers/modem/ublox-sara-r4.c
index 669c87926a..42ccef8336 100644
--- a/drivers/modem/ublox-sara-r4.c
+++ b/drivers/modem/ublox-sara-r4.c
@@ -1439,7 +1439,7 @@ static int modem_init(struct device *dev)
        ARG_UNUSED(dev);
 
        /* check for valid pinconfig */
-       __ASSERT(sizeof(pinconfig) == MAX_MDM_CONTROL_PINS,
+       __ASSERT(ARRAY_SIZE(pinconfig) == MAX_MDM_CONTROL_PINS,
               "Incorrect modem pinconfig!");
 
        (void)memset(&ictx, 0, sizeof(ictx));
@@ -1712,9 +1712,10 @@ static int offload_sendto(struct net_pkt *pkt,
        ret = send_data(sock, dst_addr, dst_port, pkt);
        if (ret < 0) {
                LOG_ERR("send_data error: %d", ret);
+       } else {
+               net_pkt_unref(pkt);
        }
 
-       net_pkt_unref(pkt);
        if (cb) {
                cb(context, ret, user_data);
        }
```

weinholtendian commented on drivers/modem/ublox-sara-r4.c:
```diff
@@ -0,0 +1,1854 @@
+			     struct sockaddr *src)
+{
+	int hdr_len = 0;
+	u16_t src_port = 0U;
```
Keeping src_port as 0 causes DTLS packets to be dropped by `dtls_is_peer_addr_valid`. I managed to get a DTLS connection established and exchanging appdata with `src_port = net_sin(&sock->dst)->sin_port;` in the AF_INET case near line 486. Note the reversal of src/dst since these packets come from dst. I imagine that the AF_INET6 case is similar, but I currently can't test that one.

mike-scott replied:
@weinholtendian: I'll take a look at this.. I didn't test DTLS yet on the modem.

mike-scott committed:
drivers: modem: wncm14a2a: dont free packet on error

Caller will handle freeing packet if error is encountered.

Fixes issue reported by Github User @weinholtendian:
<err> net_pkt: *** ERROR *** pkt 0x20027e78 is freed already
by offload_sendto():1717 (context_sendto():1473).

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: wncm14a2a: correct pinconfig assert check

Assert is checking the array size of pinconfig.  Not the actual
size of the structure.

Fixes issue reported by Github User @weinholtendian

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
dts-binding: modem: wnc-14a2a: add label binding

The label binding was missed during initial development of the
WNC-M14A2A modem.  Let's add the binding so that during the
migration to a shield, we can update the DTS correctly.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
dts/bindings: introduce Arduino-header-R3 GPIO binding

This allows HW with compatible headers to define the related GPIOs

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
boards: arm: frdm_k64f: add Arduino header definition to DTS

Let's expose the Arduino R3-compatible pin definition as DTS
for the frdm_k64f board.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
boards: nrf52840_pca10056: add Arduino header definition to DTS

Let's expose the Arduino R3-compatible pin definition as DTS
for the nrf542840_pca10056 board.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
board: disco_l475_iot1: add Arduino header definition to DTS

Let's expose the Arduino R3-compatible pin definition as DTS
for the disco_l475_iot1 board.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
samples: lwm2m_client: cleanup shield overlay files

Let's remove the sample-specific overlay files dealing with the
WNC-M14A2A modem.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
cmake: add handling for shield config and shield/board overlays

Shields can often be very complex to setup in a generic way for several
boards to support.  Let's allow shields to have their own .conf files
as well as specialized overlays per board (when needed).

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
shields: introduce shield-based WNC-M14A2A modem setup

This shield uses a non-standard UART exposed via Arduino-R3
compatible header pins.

It has configurations for FRDM_K64F and nRF52840_PCA10056.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: wncm14a2a: remove HW specific Kconfig settings

The WNC-M14A2A shield configuration has HW specific settings in place.
We can remove those settings here.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
boards: arm: frdm_k64f: don't use ETH if MODEM is enabled

Let's change the specific WNC-M14A2A check into a more generic
MODEM check for enabling ethernet.  Many of these pins are used
on the Arduino headers.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
samples: lwm2m_client: re-add sanity check for WNC-M14A2A build

Instead of using the now removed overlay files, use the SHIELD=
command-line option for testing the WNC-M14A2A modem.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
dts: bindings: add modem binding for u-blox SARA-R4

The SARA-R4 series modules from U-Blox are size-optimized LTE-M /
NB-IoT and EGPRS modules designed for low power consumption and
longer battery life.

The binding identifies the UART device, power GPIO and reset GPIO
lines.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
drivers: modem: introduce u-blox SARA-R4 modem driver

The u-blox SARA-R4 modem modules are Ultra-compact LTE Cat
M1 / NB1 ready:

- Configurable with a single hardware version
- Flexible mode selection as LTE Cat M1, LTE Cat NB1, EGPRS -
  only/preferred
- Low power consumption and longer battery life
- Extended range in buildings, basements, and with NB1,
  underground

This driver introduces support for basic AT commands to
query modem information as well as socket connection
for TCP/UDP communication.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
boards: arm: frdm_k64f: add pinmux for SARA-R4 modem

Setup the SARA-R4 POWER and RESET pins as gpios.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
boards: shields: introduce SparkFun u-blox SARA-R4 shield

SparkFun offers an Arduino-R3 compatible shield using the SARA-R410M-02B
LTE Cat M1/NB-IoT modem.

Now, that the basic SARA-R4 modem driver is implemented, let's enable
the shield for several MCUs supporting Arduino-R3 compatible headers.

Product Link:
https://www.sparkfun.com/products/14997

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott committed:
boards: arm: particle_boron: enable SARA-R4 modem

The Particle.io Boron is an nRF52840-based board with a connected
u-blox SARA-R4 modem.  The main board was previously upstreamed
without modem support.

Now that we have a driver to support the SARA-R4 modem, let's enable
it for the Boron board.

Signed-off-by: Michael Scott <mike@foundries.io>

mike-scott commented:
Rebasing this PR w/ @erwango newly merged support for gpio-map

mike-scott commented:
@weinholtendian I've integrated your previous changes (other than DTLS related) into the SARA_R4 driver.
@erwango please look at the Arduino Header setups as they are very close to what you used in the other PR, but modified slightly
@MaureenHelm I think we're in good shape with the gpio-map changes and using this PR as an example to setup MCR20A as a shield.  Please take a look again at the "cmake" patch in this series which enables this.

WilliamGFish commented:
Just pulled the latest changes and the modem that was working now just does this after completing connection to mobile network:
It may have something to do with the recent socket changes?


> 00000004 0a 4f 4b 0d 0a                                    [.OK..]
> ***** MPU FAULT *****
>   Instruction Access Violation
> ***** Hardware exception *****
> Current thread ID = 0x200048e0
> Faulting instruction address = 0x200048e0
> Fatal fault in essential thread! Spinning...

mike-scott commented:
@WilliamGFish There are a couple more changes coming, but which modem were you testing?

WilliamGFish commented:
> @WilliamGFish There are a couple more changes coming, but which modem were you testing?

I have been working with a modified sara-r4 driver on u201. The error occurs when attempting :
	/* Set iface up */
	net_if_up(ictx.iface);

erwango commented on boards/shields/sparkfun_sara_r4/boards/disco_l475_iot1.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_UART_4=y
```
@mike-scott , why not applying @MaureenHelm's proposal (in https://github.com/zephyrproject-rtos/zephyr/pull/14057/commits/d5eda49fe83f7e53c8af7926eafdda025786c5fe#r263920854) here?

This is close to what we should get in future, if get read of Kconfig symbols per device instance. Then, the following would do the same:
```
&arduino_spi {
               status = "ok";
}
```

mike-scott replied:
@erwango This PR is now suffering from "scope creep".  There are several more additions I want to do, but felt like it would be best to land it as is and then make this changes.

mike-scott replied:
Also, this approach doesn't work for WNC-M14A2A (the original modem driver in this PR) -- it uses non-standard pins for UART on the Arduino headers.

erwango replied:
> @erwango This PR is now suffering from "scope creep". 

I'd rather say this is an ambitious PR that could have been split in small pieces.
Please not that I retained myself to request again for splitting board connector dts definitions into dedicated .dtsi files :-).

Anyway, agreed to let this as a further enhancement.

erwango commented on dts/bindings/modem/ublox,sara-r4.yaml:
```diff
@@ -0,0 +1,37 @@
+#
```
You'll need a commit message in commit header

mike-scott replied:
Ah yes, I'll fix.  Thanks.

mike-scott commented:
> > @WilliamGFish There are a couple more changes coming, but which modem were you testing?
> 
> I have been working with a modified sara-r4 driver on u201 on a particle boron.
> Up until I pulled the last ~13 commits it had been working just fine, odd errors but there network related.
> 
> The error occurs when attempting :
> /* Set iface up */
> net_if_up(ictx.iface);

Can  you try testing w/ these 2 changes to the prj.conf:
CONFIG_NET_IPV6=n
CONFIG_NET_CONFIG_NEED_IPV6=n

mike-scott commented:
Rebased and addressed gitlint and license issues.
